Call implementations of the Fn* traits on references

mikem8891 opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

This pull request fixes [Issue #42736](https://github.com/rust-lang/rust/issues/42736), allowing references of `struct`s, `enum`s, and `union`s to implement `Fn*` traits and be called.  For example, the code below will compile and run properly.  If approved this should also push [Issue #29625](https://github.com/rust-lang/rust/issues/29625) forward as well.

```
#![feature(fn_traits)]
#![feature(unboxed_closures)]

struct S;

impl FnOnce<()> for &S {
    type Output = ();

    extern "rust-call" fn call_once(self, _arg: ()) {
        println!("Haldo, world!");
    }
}

fn main() {
    let s = S;
    (&s)();
}
```

mikem8891 committed:
Update callee.rs

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13404237924/job/37441108390) [(plain)](https://github.com/rust-lang/rust/commit/4c56758499e5528ad1624894911c9150b66d66a5/checks/37441108390/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04
ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
  make \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 2.753 Building wheels for collected packages: reuse
#12 2.754   Building wheel for reuse (pyproject.toml): started
#12 2.969   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 2.970   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=0c2fd2aaf7b0bf8d6e131220aff14712a774c2ca462f3204d25460cbcf610b63
#12 2.970   Stored in directory: /tmp/pip-ephem-wheel-cache-20z984su/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 2.973 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.379 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.379 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 3.914 Collecting virtualenv
#12 3.914 Collecting virtualenv
#12 3.999   Downloading virtualenv-20.29.2-py3-none-any.whl (4.3 MB)
#12 4.134      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 32.9 MB/s eta 0:00:00
#12 4.191 Collecting filelock<4,>=3.12.2
#12 4.211   Downloading filelock-3.17.0-py3-none-any.whl (16 kB)
#12 4.230 Collecting distlib<1,>=0.3.7
#12 4.249   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.256      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 97.0 MB/s eta 0:00:00
#12 4.288 Collecting platformdirs<5,>=3.9.1
#12 4.308   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#12 4.391 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.574 Successfully installed distlib-0.3.9 filelock-3.17.0 platformdirs-4.3.6 virtualenv-20.29.2
#12 DONE 4.7s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      124864 kB
DirectMap2M:     8263680 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2025-02-08/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
##[endgroup]
fmt check
Diff in /checkout/compiler/rustc_hir_typeck/src/callee.rs:224:
             //
             // The simplest fix by far is to just ignore this case and deref again,
             // so we wind up with `FnMut::call_mut(&mut *f, ())`.
-            ty::Ref(_, deref_ty, _) if autoderef.step_count() == 0 => {
-                    ty::Adt(..) => {},
-                    ty::Adt(..) => {},
-                    _ => {return None}
-            }
-            }
+            ty::Ref(_, deref_ty, _) if autoderef.step_count() == 0 => match deref_ty.kind() {
+                _ => return None,
+            },
 
             ty::Error(_) => {
```

</details>

workingjubilee commented:
huh, meaning-changing formatting...?

theemathas commented:
@workingjubilee It's not a "meaning-changing formatting" issue. For some reason, the diff in rust-lig-analyzer's comment does not match the diff in the actual logs.

workingjubilee commented:
oh weird.

compiler-errors commented:
The diff in rust-log-analyzer just renders weirdly for some reason.

However, this still needs a test associated with it, and a pretty clear explanation of the root cause of the issue and why this fixes it. Also, yeah, please fix the formatting.

I also don't think we should land this regardless, personally; `Fn*` traits are pretty internal still, and there's specific HIR typeck hacks that need to be adjusted in general.


@rustbot author

mikem8891 committed:
Update callee.rs

Added comments and improved formatting.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13411671142/job/37463088387) [(plain)](https://github.com/rust-lang/rust/commit/d97eaf80abf2fc88abfe09bd54d1f5669f4d42bd/checks/37463088387/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#21 exporting to docker image format
#21 sending tarball 27.2s done
#21 DONE 40.8s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
```

</details>

mikem8891 commented:
I fixed the formatting and added a short explanation of the change in the comments.

As for the original cause of the (Issue #42736)[https://github.com/rust-lang/rust/issues/42736], it is caused during the probing phase of type checking.  `rustc_hir_typeck::callee::FnCtxt::try_overloaded_call_step` contains a hack that causes all `Ref` Candidates to be automatically discarded so the `autoderef` can move on to the next  step of dereferencing.  This allows `&fn()` and similar closures to be dereferenced further.  This allows references to closures and function pointers that implement and call using `FnMut` to not require a `&mut`.  This should not apply to structs, enums, or unions; as they would actually mutate `self`.

The tests to add seem like they would be pretty straight forward.  This is my first pull request, so I apologize if I have missed any steps.  Where in the `tests` should I put my tests? 

Unless I am actively working against someone who is already working to fix this and the other HIR typeck hacks, I don't see any harm in fixing this issue now.

Thank you!
@rustbot review

compiler-errors commented:
>  Unless I am actively working against someone who is already working to fix this and the other HIR typeck hacks, I don't see any harm in fixing this issue now.

This adds complexity to HIR typeck for a compiler-only feature. I don't think we really should be fixing this, especially because I cannot see an `FnOnce for &Adt` being implemented in a practical way without a corresponding `Fn` impl.

Thanks for the contribution, but I'm gonna opt to close it.

compiler-errors commented:
> If approved this should also push https://github.com/rust-lang/rust/issues/29625 forward as well.

The blocker for stabilizing the `Fn` traits is a lot more than just trivial implementation bugs; there's a lot of open questions and caveats about exposing the `extern "rust-call"` ABI and how to represent function arguments and variadics and other things like that.
